README.md
> Wer vorab tetsen möchte kann das gerne wie folgt machen: 
> Einige Sachen funktionieren noch nicht oder sind noch nicht getestet.
> Issues/Wünsche könnt Ihr trotzdem schon schreiben :-)


> Sollte das Backend (viel) zu langsam sein bitte in der .yml Datei **compile:** auf 0 stellen.


## Das Modul // Codename: _Gensfleisch 1468_ 

Dieses AddOn installiert das Modul **0000 - Standard** und die zugehörigen Dateien.

Auf einer Einstellungsseite im Modul kann die jeweile Anordung der Inhalte ausgewählt werden (jederzeit änderbar).
Hier gibt es auch die Möglichkeit für den Container, der Row und den Cols IDs und Klassen zu vergeben.
Es besteht zudem die Möglichkeit zu wählen ob der Container fluid sein soll.


---

### Installation / Benutzung

folgt...


---
### Module (eigentlich Funktionen)

**0010 - Überschrift** (headline_input / headline_output) 
Hier kann eine Überschrift eingepflegt werden. Zusätzlich ist die Angabe der "Größe" (H1-H6) ist möglich.

---

**0020 - Text** (textarea_input / textarea_output)
Es wird eine Textarea bereitgestellt. Je nach installiertem Editor wird dieser eingebunden. Aktuell funktionieren leider nur die AddOns: _MarkitUp_ und _Tinymce4_ (siehe: https://github.com/FriendsOfREDAXO/redactor2/issues/134)<br/>_Sollten schon Inhalte eingepflegt sein können bei einem Wechsel des Editors alle Formtierungen verloren gehen._

---

**0030 - Bild** (image_input / image_output)
Hier kann ein Bild ausgewählt und intern verlinkt werden.
*CSS: ./assets/addons/das_modul/css/image.css*  

---

**0040 - Download** (downloads_input / downloads_output
Dateien können zum Downlad bereitgestellt werden.

Wem der Download Pfad nicht gefällt, der wandelt die .htaccess noch ab:

    RewriteRule ^media/([^/]*)/([^/]*) %{ENV:BASE}/index.php?rex_media_type=$1&rex_media_file=$2&%{QUERY_STRING} [B]

Der Download erfolgt dann über /media/download/dateiname.pdf</p>

Oder:

    RewriteRule ^download/([^/]*) %{ENV:BASE}/index.php?rex_media_type=download&rex_media_file=$1&%{QUERY_STRING} [B]

Der Download erfolgt dann direkt über /download/dateiname.pdf

(Nochmal testen)

--- 
 
**0050 - Film (extern)** (video_input / video_output)
Durch die Angabe eine YouTube bz. Vimeo Film ID kann das Video im Fornten dargestellt werden. Hier sollte noch eine DSGVO gerechte Lösung gefunden werden.
*CSS: ./assets/addons/das_modul/css/video.css*

---

**0060 - Link (intern / extern)** (link_input / link_output)

---

**0070 - Card (Teaser)**
Ermöglicht es eine CARD auszugeben (Bild, Text, interner Link). *CSS: ./assets/addons/das_modul/css/cards.css*  

---

**0080 - Abstand einfügen**
Es kann ein abstand mir/ohne Grafik/line ausgegeben werden.
*Die Grafiuk liegt hier: ./assets/addons/das_modul/images/divider.png* 

---

**0090 - Artikel im Modal öffnen**

---

**0100 - Gallery / Carousel**
Einbindung der Unite Gallery


---

### Sonstige Funktionen

- **check Editor**
Hier wird geprüft welcher Editor installiert ist.<br/>
MarkitUp   - funktioniert
Tinymce4   - funktioniert
CKE5       - funktioniert
Redactor 2 - [funktioniert nicht](https://github.com/FriendsOfREDAXO/redactor2/issues/134)

---

- **container_input**
Funktion um die "Breite" des Containers zu definieren. 

---

- **id_class_input**
Funktion für die Eingaben der Klassen und IDs

---

> **Warum _Gensfleisch 1468_**
>
> Nun WordPress bekommt jetzt demnächst den ["Gutenberg" Editor](https://de.wordpress.org/gutenberg/) und die freuen sich grad ´n Ast. Das was der supertolle Gutenberg Editor kann/können wird ist schon lange die Funktionsweise von REDAXO. Vielleicht nicht ganz so fancy. Dieses Modul ermöglicht es dem Redakteur Inhalte modular und im Grid wie beim Gutenberg Editor (geplant) zu pflegen. Nur einfacher :-)  
>
> Gutenberg heisst eigentlich "Johannes Gensfleisch" und ist 1468 gestorben... (https://de.wikipedia.org/wiki/Johannes_Gutenberg)


---
<br/>

Test Template:

```<!doctype html>
   <html lang="en">
   <head>
       <meta charset="utf-8">
       <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
   
       <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css">
   
       <link rel="stylesheet" href="./assets/addons/das_modul/css/image.css">
       <link rel="stylesheet" href="./assets/addons/das_modul/css/video.css">
       <link rel="stylesheet" href="./assets/addons/das_modul/css/cards.css">
       <link rel="stylesheet" href="./assets/addons/das_modul/css/abstand.css">
       <link rel="stylesheet" href="./assets/addons/das_modul/css/modal.css">
       <!-- <link rel="stylesheet" href="./assets/addons/das_modul/css/unitegallerycss.css"> -->
   
   
       <script type='text/javascript' src='./assets/addons/das_modul/vendor/unitegallery/js/jquery-11.0.min.js'></script>
       <script type='text/javascript' src='./assets/addons/das_modul/vendor/unitegallery/js/unitegallery.min.js'></script>
   
       <link rel='stylesheet' href='./assets/addons/das_modul/vendor/unitegallery/css/unite-gallery.css' type='text/css' />
   
       <script type='text/javascript' src='./assets/addons/das_modul/vendor/unitegallery/themes/default/ug-theme-default.js'></script>
   
   
       <script type='text/javascript' src='./assets/addons/das_modul/vendor/unitegallery/themes/carousel/ug-theme-carousel.js'></script>
       <script type='text/javascript' src='./assets/addons/das_modul/vendor/unitegallery/themes/compact/ug-theme-compact.js'></script>
   
   
       <script type='text/javascript' src='./assets/addons/das_modul/vendor/unitegallery/themes/tiles/ug-theme-tiles.js'></script>
       <script type='text/javascript' src='./assets/addons/das_modul/vendor/unitegallery/themes/tilesgrid/ug-theme-tilesgrid.js'></script>
       <script type='text/javascript' src='./assets/addons/das_modul/vendor/unitegallery/themes/grid/ug-theme-grid.js'></script>
   
       <script type='text/javascript' src='./assets/addons/das_modul/vendor/unitegallery/themes/slider/ug-theme-slider.js'></script>
   
   
       <link rel='stylesheet' 		  href='./assets/addons/das_modul/vendor/unitegallery/themes/default/ug-theme-default.css' type='text/css' />
   
   
   </head>
   <body>
   
   REX_ARTICLE[]
   
   
   
   
   <script>
       $(".open-rex_modal").on('click', function(e){
   
           $(".close-rex_modal").click();
   
   
           e.preventDefault();
           e.stopImmediatePropagation;
   
           var $this = $(this),
               modal = $($this).data("modal");
   
   
           $(modal).parents(".rex_modal-overlay").addClass("open");
           $(modal).addClass("open");
   
           // $(modal).parents(".rex_modal-overlay").insertBefore( "#footer" );
   
           $(document).on('click', function(e){
               var target = $(e.target);
   
               if ($(target).hasClass("rex_modal-overlay")){
                   $(target).find(".rex_modal").each( function(){
                       $(this).removeClass("open");
                   });
                   $(target).removeClass("open");
               }
           });
       });
   
       $(".close-rex_modal").on('click', function(e){
           e.preventDefault();
           e.stopImmediatePropagation;
   
           var $this = $(this),
               modal = $($this).data("modal");
   
           $(modal).removeClass("open");
           $(".rex_modal-overlay").removeClass("open");
   
       });
   
       function printDiv(divId) {
           var content = document.getElementById(divId).innerHTML;
           var mywindow = window.open('', 'Print', 'height=10,width=10');
   
           mywindow.document.write('<html><head><title>Print</title>');
           mywindow.document.write('</head><body>');
           mywindow.document.write(content);
           mywindow.document.write('</body></html>');
   
           mywindow.document.close();
           mywindow.focus()
           mywindow.print();
           mywindow.close();
           return true;
       }
   </script>
   
   </body>
   </html>
```




___

### Credits

- [REDAXO](https://redaxo.org), [FriendsOfREDAXO](https://github.com/FriendsOfREDAXO)
- [MForm](https://github.com/FriendsOfREDAXO/mform) , [MBlock](https://github.com/FriendsOfREDAXO/mblock)
- [Bootstrap](https://getbootstrap.com/), [Unite Gallery](https://github.com/vvvmax/unitegallery/) 
- [Tim](https://github.com/orgs/FriendsOfREDAXO/people/elricco),[Thomas](https://github.com/orgs/FriendsOfREDAXO/people/tbaddade),[Peter](https://github.com/polarpixel),[Gregor](https://github.com/orgs/FriendsOfREDAXO/people/gharlan) (und alle, die hier vergessen wurden :-))

CHANGELOG.md
Das Modul - Changelog
=====================

Version X  – 07.09.2018
-----------------------

* Alpa


boot.php
<?php

$this->setProperty('author', 'concedra GmbH / Oliver Kreischer');

$this->setProperty('stylescss' ,'0');
$this->setProperty('videocss'  ,'0');
$this->setProperty('abstandcss','0');
$this->setProperty('cardscss'  ,'0');
$this->setProperty('modalcss'  ,'1');
$this->setProperty('unitegallerycss'  ,'0');

if (rex::isBackend() && rex::getUser()) {

    rex_perm::register('das_modul[]');

    rex_extension::register('PACKAGES_INCLUDED', function () {
        if (rex::getUser() && $this->getProperty('compile') && $this->getProperty('stylescss')) {

            $compiler = new rex_scss_compiler();
            $scss_files = rex_extension::registerPoint(new rex_extension_point('BE_STYLE_SCSS_FILES', [$this->getPath('scss/master.scss')]));
            $compiler->setScssFile($scss_files);
            $compiler->setCssFile($this->getPath('assets/css/styles.css'));
            $compiler->compile();
            rex_file::copy($this->getPath('assets/css/styles.css'), $this->getAssetsPath('css/styles.css'));
        }
    });
    rex_view::addCssFile($this->getAssetsUrl('css/styles.css'));


    /** Image **/
    rex_extension::register('PACKAGES_INCLUDED', function () {
        if (rex::getUser() && $this->getProperty('compile') && $this->getProperty('imagecss')) {
            $compiler = new rex_scss_compiler();
            $scss_files = rex_extension::registerPoint(new rex_extension_point('BE_STYLE_SCSS_FILES', [$this->getPath('scss/image.scss')]));
            $compiler->setScssFile($scss_files);
            $compiler->setCssFile($this->getPath('assets/css/image.css'));
            $compiler->compile();
            rex_file::copy($this->getPath('assets/css/image.css'), $this->getAssetsPath('css/image.css'));
        }
    });

    /** Video **/
    rex_extension::register('PACKAGES_INCLUDED', function () {
        if (rex::getUser() && $this->getProperty('compile') && $this->getProperty('videocss')) {
            $compiler = new rex_scss_compiler();
            $scss_files = rex_extension::registerPoint(new rex_extension_point('BE_STYLE_SCSS_FILES', [$this->getPath('scss/video.scss')]));
            $compiler->setScssFile($scss_files);
            $compiler->setCssFile($this->getPath('assets/css/video.css'));
            $compiler->compile();
            rex_file::copy($this->getPath('assets/css/video.css'), $this->getAssetsPath('css/video.css'));
        }
    });

    /** Abstand **/
    rex_extension::register('PACKAGES_INCLUDED', function () {
        if (rex::getUser() && $this->getProperty('compile') && $this->getProperty('abstandcss')) {

            $compiler = new rex_scss_compiler();
            $scss_files = rex_extension::registerPoint(new rex_extension_point('BE_STYLE_SCSS_FILES', [$this->getPath('scss/abstand.scss')]));
            $compiler->setScssFile($scss_files);
            $compiler->setCssFile($this->getPath('assets/css/abstand.css'));
            $compiler->compile();
            rex_file::copy($this->getPath('assets/css/abstand.css'), $this->getAssetsPath('css/abstand.css'));
        }
    });

    /** Cards **/
    rex_extension::register('PACKAGES_INCLUDED', function () {
        if (rex::getUser() && $this->getProperty('compile') && $this->getProperty('cardscss')) {

            $compiler = new rex_scss_compiler();
            $scss_files = rex_extension::registerPoint(new rex_extension_point('BE_STYLE_SCSS_FILES', [$this->getPath('scss/cards.scss')]));
            $compiler->setScssFile($scss_files);
            $compiler->setCssFile($this->getPath('assets/css/cards.css'));
            $compiler->compile();
            rex_file::copy($this->getPath('assets/css/cards.css'), $this->getAssetsPath('css/cards.css'));
        }
    });

    /** Modal **/
    rex_extension::register('PACKAGES_INCLUDED', function () {
        if (rex::getUser() && $this->getProperty('compile') && $this->getProperty('modalcss')) {

            $compiler = new rex_scss_compiler();
            $scss_files = rex_extension::registerPoint(new rex_extension_point('BE_STYLE_SCSS_FILES', [$this->getPath('scss/modal.scss')]));
            $compiler->setScssFile($scss_files);
            $compiler->setCssFile($this->getPath('assets/css/modal.css'));
            $compiler->compile();
            rex_file::copy($this->getPath('assets/css/modal.css'), $this->getAssetsPath('css/modal.css'));
        }
    });

    /** Unite Gallery **/
    rex_extension::register('PACKAGES_INCLUDED', function () {
        if (rex::getUser() && $this->getProperty('compile') && $this->getProperty('unitegallerycss')) {
            $compiler = new rex_scss_compiler();
            $scss_files = rex_extension::registerPoint(new rex_extension_point('BE_STYLE_SCSS_FILES', [$this->getPath('scss/unitegallerycss.scss')]));
            $compiler->setScssFile($scss_files);
            $compiler->setCssFile($this->getPath('assets/css/unitegallerycss.css'));
            $compiler->compile();
            rex_file::copy($this->getPath('assets/css/unitegallerycss.css'), $this->getAssetsPath('css/unitegallerycss.css'));
        }
    });


}




lib/modul/input.php
<?php

/*

* Überschrift mit Anker


*/


$values         = array();
$values[1]      = array( '1', rex_var::toArray( 'REX_VALUE[1]' ) );
$values[2]      = array( '2', rex_var::toArray( 'REX_VALUE[2]' ) );
$values[3]      = array( '3', rex_var::toArray( 'REX_VALUE[3]' ) );
$values[4]      = array( '4', rex_var::toArray( 'REX_VALUE[4]' ) );
$wrapper[5]     = array( '5', rex_var::toArray( 'REX_VALUE[5]' ) );
$counter        = 0;
$reihenfolgeneu = '';

$dm_input = NEW rex_das_modul_helper();


echo '
  <div id="das_modul_wrapper">
    <input id="anrodnung" type="hidden" name="REX_INPUT_VALUE[19]" value="REX_VALUE[19]" />
    <div id="tabs">
      <ul class="nav nav-tabs">';
if ( 'REX_VALUE[19]' ) {
    $reihenfolgeneu = explode( ',', 'REX_VALUE[19]' );
    for ( $i = 1; $i <= count( $values ); $i ++ ) {
        echo '<li id="' . ( $reihenfolgeneu[ ( $i - 1 ) ] ) . '" ><a data-toggle="tab" id="tab' . ( $reihenfolgeneu[ ( $i - 1 ) ] ) . '" href="#bereich' . ( $reihenfolgeneu[ ( $i - 1 ) ] ) . '"><i>B' . $i . '</i><span>Bereich ' . $i . '</span></a></li>';
    }
} else {
    for ( $i = 1; $i <= count( $values ); $i ++ ) {
        echo '<li id="' . $i . '"><a data-toggle="tab" id="tab' . $i . '" href="#bereich' . $i . '"><i>B' . $i . '</i><span>Bereich ' . $i . '</span></a></li>';
    }
}
echo '  <li class="locked pull-right"><a data-toggle="tab" id="tabweiteres" href="#einstellungen"><i class="rex-icon rex-icon-metafuncs"></i><span>Einstellungen</span></a></li>
      </ul>
  </div>';
echo '<div class="tab-content">';
for ( $i = 1; $i <= count( $values ); $i ++ ) {
    echo '<div id="bereich' . $i . '" class="tab-pane fade in">';
    $id    = $i;
    $mform = new MForm();
    $mform->addFieldset( 'Element', array( 'class' => 'elemente' ) );
    $mform->addSelectField( "$id.0.element", array( 'empty'      => 'Element auswählen',
        'headline'      => '0010 - Überschrift',
        'textarea'      => '0020 - Text',
        'image'         => '0030 - Bild',
        'downloads'     => '0040 - Downloads',
        'video'         => '0050 - Film (extern)',
        'link'          => '0060 - Link (intern / extern)',
        'card'          => '0070 - Card (Teaser)',
        'space'         => '0080 - Abstand einfügen',
        'artikel_modal' => '0090 - Artikel im Modal öffnen',
        'unite_gallery' => '0100 - Gallery / Carousel'

    ), array( 'label' => '', 'class' => 'elementSelect' ) );


    $dm_input->headline_input(  $id, $mform );
    $dm_input->textarea_input(  $id, $mform );
    $dm_input->image_input(     $id, $mform );
    $dm_input->downloads_input( $id, $mform );
    $dm_input->video_input(     $id, $mform );
    $dm_input->link_input(      $id, $mform );
    $dm_input->card_input(      $id, $mform );
    $dm_input->space_input(     $id, $mform );
    $dm_input->modal_input(     $id, $mform );
    $dm_input->unite_gallery(   $id, $mform );

    echo MBlock::show( $id, $mform->show() );
    echo '<div class="extra_settings">Weitere Einstellungen<i class="fa fa-arrow-down" aria-hidden="true"></i></div>
        <div class="extra_settings_content">';

    $dm_input->id_class_input( $id );

    echo '</div>
  </div>';
}
$options = array(
    '12'      => '<div class="img12"></div>',
    '6_6'     => '<div class="img6_6"></div>',
    '4_4_4'   => '<div class="img4_4_4"></div>',
    '3_3_3_3' => '<div class="img3_3_3_3"></div>',
    '6_3_3'   => '<div class="img6_3_3"></div>',
    '3_6_3'   => '<div class="img3_6_3"></div>',
    '3_3_6'   => '<div class="img3_3_6"></div>',
    '8_4'     => '<div class="img8_4"></div>',
    '4_8'     => '<div class="img4_8"></div>'
);
$grid    = '';
foreach ( $options as $value => $label ) {
    $checked = 'REX_VALUE[20]' == $value ? ' checked="checked"' : '';
    $grid    .= '<label><input name="REX_INPUT_VALUE[20]" value="' . $value . '" type="radio"' . $checked . ' />' . $label . '</label>';
}
echo '
    <div id="einstellungen" class="tab-pane fade in active">
      <div class="mform">
        <fieldset class="form-horizontal">
          <legend>Grid Auswahl <i class="module_help_link fa fa-exclamation-triangle" aria-hidden="true"></i></legend>
              <div class="module_help_content">
                <p>Sofern Sie das Raster nachträglich ändern möchten ist es notwendig den Block vor der Änderung einmal zu speichern (ein Klick auf "Block übernehmen" reicht). Dies ist nur wichtig sofern Sie die Anordnung der Tabs vorher geändert haben.</p>
              </div>
          <div class="col-sm-6 col-md-5 col-lg-5">
            <p>Wie soll der Inhalt dargestellt werden?</p>
            <p>Die Auwahl des Rasters kann jederzeit einfach geändert werden. Bitte beachten Sie hierzu den Hinweis!</p>
          </div>
          <div class="col-sm-6 col-md-7 col-lg-7 grid">
            ' . $grid . '
          </div>
        </fieldset>
      </div>';
$dm_input->container_input( 5 );
$dm_input->id_class_input( 5 );
// $dm_input->media_manager_typ_input( 5 );
echo '
  </div>
</div>';
?>
<script>
    $(function () {
        $(document).on('change', '.elementSelect', function () {
            //console.log($(this).val());
            //console.log($(this).parents('.sortitem').find('fieldset.' + $(this).val()));
            $($(this).children()).each(function () {
                $(this).parents('.sortitem').find('fieldset.' + $(this).val()).hide();
            });
            $(this).parents('.sortitem').find('fieldset.' + $(this).val()).toggle();
        });

        $(document).on('mblock:add', function (event, container) {
            $('.elementSelect option:selected').each(function (index) {
                // console.log($(this).val());
                if ($(this).val() == 'empty') {
                    $(this).parents('.sortitem').find('.elementSelect option').each(function (index) {
                        $(this).parents('.sortitem').find('fieldset.' + $(this).val()).hide();
                    });
                } else {
                    $(this).val() == 'empty';
                }
            });
        });

        //only show selected elements on edit
        $('.elementSelect option:selected').each(function (index) {
            $(this).parents('.sortitem').find('fieldset.' + $(this).val()).show();
        });

        $(".module_help_link").click(function () {
            $(this).parent().parent().find(".module_help_content").slideToggle(100);
        });
        $(".extra_settings").click(function () {
            $(this).next().slideToggle(200);
        });
        $(function () {
            $("#tabs ul.nav-tabs ").sortable({
                axis: "x",
                items: '> li:not(.locked)',
                update: function (e, ui) {
                    var csv = "";
                    $("#tabs > ul > li").each(function (i) {
                        csv += ( csv == "" ? "" : "," ) + this.id;
                    });
                    // alert (csv);
                    $("#anrodnung").val(csv);
                }
            });
        });

        if ('REX_VALUE[20]' == '') {
            $('a#tabweiteres').click();
        } else {
            $('#tabs ul.nav-tabs li a').first().click();
        }

        function grid(str) {
            if (str == '12') {
                $('#tab1').css('display', 'block');
                $('#tab2').css('display', 'none');
                $('#tab3').css('display', 'none');
                $('#tab4').css('display', 'none');
            } else if (str == '6_6' || str == '8_4' || str == '4_8') {
                $('#tab1').css('display', 'block');
                $('#tab2').css('display', 'block');
                $('#tab3').css('display', 'none');
                $('#tab4').css('display', 'none');
            } else if (str == '4_4_4' || str == '6_3_3' || str == '3_6_3' || str == '3_3_6') {
                $('#tab1').css('display', 'block');
                $('#tab2').css('display', 'block');
                $('#tab3').css('display', 'block');
                $('#tab4').css('display', 'none');
            } else {
                $('#tab1').css('display', 'block');
                $('#tab2').css('display', 'block');
                $('#tab3').css('display', 'block');
                $('#tab4').css('display', 'block');
            }
        }
        grid('REX_VALUE[20]');
        $('.grid input[type=radio]').change(function () {
            $('#tab1').click();
            grid(this.value);
        });
    });
</script>
